### PR TITLE
Margaret 1801 final

### PR DESCRIPTION
**Pull request by lilwillifo**
This went well! I had some minor issues along the way, but was generally able to solve. Huzzah for many to many! I had to setup database cleaner to keep testing for the same movies/genres. 

One thing that is bothering me - in my genres index I have a line <% if current_user && current_user.admin? %>. I thought should be able to access the current_admin? method, but I kept getting a no method error there.

### CODE REVIEW

File: app/controllers/genres_controller.rb [hunk 1]
```diff
@@ -0,0 +1,33 @@
+class GenresController < ApplicationController
+  before_action :require_admin, only: %i[create]
```
Comment (iandouglas) [inline app/controllers/genres_controller.rb]: nice!

File: app/controllers/genres_controller.rb [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+  def index
+    @genres = Genre.all
+    @genre = Genre.new
+  end
+
+  def create
+    genre = Genre.new(genre_params)
+    if genre.save
+      redirect_to genres_path
```
Comment (iandouglas) [inline app/controllers/genres_controller.rb]: aw, no fancy flash message if it worked?

File: app/controllers/genres_controller.rb [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+  end
+
+  private
+
+  def genre_params
+    params.require(:genre).permit(:name)
+  end
+
+  def require_admin
+    render file: '/public/404' unless current_admin?
```
Comment (iandouglas) [inline app/controllers/genres_controller.rb]: nice

File: app/models/genre.rb [hunk 4]
```diff
@@ -0,0 +1,8 @@
+class Genre < ApplicationRecord
+  has_many :genre_movies
+  has_many :movies, through: :genre_movies
+
+  def average_rating
+    movies.average(:rating)
```
Comment (iandouglas) [inline app/models/genre.rb]: great use of the AR relationship here

File: app/views/genres/index.html.erb [hunk 5]
```diff
@@ -0,0 +1,11 @@
+<% @genres.each do |genre| %>
+  <%= link_to genre.name, genre_path(genre) %>
+<% end %>
+
+<% if current_user && current_user.admin? %>
```
Comment (iandouglas) [inline app/views/genres/index.html.erb]: Hint: You have `current_admin?` defined in ApplicationController, but it's not exposed as a method that can be used outside of other controllers.

File: app/views/genres/index.html.erb [hunk 6]
```diff
@@ -0,0 +1,11 @@
...
+  <%= form_for @genre do |f| %>
+    <%= f.label :name %>
+    <%= f.text_field :name %>
+    <%= f.submit 'Submit' %>
```
Comment (iandouglas) [inline app/views/genres/index.html.erb]: better practice is to let Ruby do its magic and populate the button text with "Create Genre" or "Edit Genre" if you reuse the form etc.
image

File: config/routes.rb [hunk 7]
```diff
@@ -10,6 +10,7 @@
   resources :actors
 
   resources :users, only: [:new, :create, :show]
+  resources :genres, only: [:index, :create, :show]
```
Comment (iandouglas) [inline config/routes.rb]: nice work only giving access to what you need!

File: db/migrate/20180417151424_create_genres.rb [hunk 8]
```diff
@@ -0,0 +1,7 @@
+class CreateGenres < ActiveRecord::Migration[5.1]
+  def change
+    create_table :genres do |t|
+      t.string :name
+    end
```
Comment (iandouglas) [inline db/migrate/20180417151424_create_genres.rb]: best practice to add timestamps on tables like this

File: db/migrate/20180417154149_create_genre_movies.rb [hunk 9]
```diff
@@ -0,0 +1,8 @@
+class CreateGenreMovies < ActiveRecord::Migration[5.1]
+  def change
+    create_table :genre_movies do |t|
+      t.references :movie, foreign_key: true
+      t.references :genre, foreign_key: true
+    end
```
Comment (iandouglas) [inline db/migrate/20180417154149_create_genre_movies.rb]: thanks for NOT adding timestamps here, it's rare that you WANT timestamps in a join table unless you absolutely need to know WHEN you made that association.

File: spec/features/admin_creates_genres_spec.rb [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+      allow_any_instance_of(ApplicationController).to receive(:current_user).and_return(admin)
+
+      visit(genres_path)
+
+      fill_in 'Name', with: 'Action'
+      click_on 'Submit'
+
+      expect(current_path).to eq(genres_path)
+      expect(page).to have_content('Action')
+    end
```
Comment (iandouglas) [inline spec/features/admin_creates_genres_spec.rb]: ruh-roh, no test for the sad-path of a genre not saving ... which would have caught the problem I mentioned earlier in your GenreController with the missing instance variable.

File: spec/features/visitor_sees_genre_show_page_spec.rb [hunk 11]
```diff
@@ -0,0 +1,24 @@
+require 'rails_helper'
+
+describe 'As a visitor' do
+  context 'when I visit a genre show page' do
+    it 'I see the genre name, all associated movies, and the avg ratings for these movies' do
+      genre = Genre.create!(name: 'Action')
+      director = Director.create!(name: 'Ian Douglas')
```
Comment (iandouglas) [inline spec/features/visitor_sees_genre_show_page_spec.rb]: hey, I know that guy!

File: spec/features/visitor_sees_genre_show_page_spec.rb [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+      visit genres_path
+
+      click_on genre.name
+
+      expect(current_path).to eq(genre_path(genre))
+      expect(page).to have_content(genre.name)
+      expect(page).to have_content(movie_1.title)
+      expect(page).to have_content(movie_2.title)
+      expect(page).to_not have_content(movie_3.title)
+      expect(page).to have_content("Average Movie Rating: #{genre.average_rating}")
```
Comment (iandouglas) [inline spec/features/visitor_sees_genre_show_page_spec.rb]: good job using string interpolation here

File: spec/features/visitor_sees_genres_index_spec.rb [hunk 13]
```diff
@@ -0,0 +1,20 @@
...
+      visit genres_path
+
+      expect(page).to_not have_button('Submit')
+      expect(page).to have_content(genre_1.name)
+      expect(page).to have_content(genre_2.name)
+
+      click_on genre_1.name
+
+      expect(current_path).to eq(genre_path(genre_1))
+    end
```
Comment (iandouglas) [inline spec/features/visitor_sees_genres_index_spec.rb]: I'd like to see a test here that ensures a non-admin does NOT see the genre creation form.

File: spec/models/genre_spec.rb [hunk 14]
```diff
@@ -0,0 +1,18 @@
+require 'rails_helper'
+
+describe Genre, type: :model do
+  describe 'relationships' do
+    it {is_expected.to have_many(:genre_movies)}
+    it {is_expected.to have_many(:movies).through(:genre_movies)}
+  end
+  describe 'instance methods' do
+    it '.average_rating' do
```
Comment (iandouglas) [inline spec/models/genre_spec.rb]: woot, model tests!!!! 😍

File: spec/features/visitor_sees_genre_show_page_spec.rb (hunk 11 above)
Comment (lilwillifo) [reply spec/features/visitor_sees_genre_show_page_spec.rb]: 👋

**Pull request closed without merging**